design/proposals/spike-cli-dockerless.md

```diff
@@ -0,0 +1,257 @@
+
+1 - the least work
+
+3 - the most work
+
+|            | Functionality | Quality | UX   | Deployment | Testing | Security |
+| ---------- | ------------- | ------- | ---- | ---------- | ------- | -------- |
+| Ruby CLI   | 1             | 1       | 2    | 3          | 1       | See scan |
+| Go CLI     | 3             | 3       | 3    | 2          | 3       | See scan |
+| Python CLI | 2             | 2       | 1    | 1          | 2       | See scan |
+
+Ruby CLI ~11 weeks
+
+- Ramp-up on creating packages and create design for CLI (~1 week)
+- Provide package (rpm, pkg, deb, etc) / [deliver as executable](https://www.ruby-toolbox.com/categories/packaging_to_executables) (CLI + dependencies) (~3 week)
+- Expand current testing if needed (~1 week)
+- Build infrastructure for testing on new platforms (Windows, macOS, REHL environments) (~3 weeks)
+- Run a few CLI commands on new supported platforms (Windows, macOS environments) (~2 weeks)
+- Close open issues (~1 week)
+
+Python CLI ~13 weeks
+
+- Python ramp-up and create design for CLI (~1 week)
+- Implement logic to fill in functionality gaps (7 CLI commands, ~3 weeks)
+  - `init`, `authn` (`login`, `authenticate`), `show `, `rotate_api_key` (`host` / `user`)
+- Write associated tests for new implementation (~3 weeks)
+- Build infrastructure for testing on new platforms (Windows, macOS environments) (~3 weeks)
+- Run a few of them on different supported platforms (Windows, macOS, REHL environments) (~2 weeks)
+- Close open issues (~1 week)
+- Add documentation
+
+Go CLI - Complete rewrite (~18+ weeks)
+
+- Create design for CLI
+- CLI Ramp up
+- Redesign CLI
+- Implement new CLI commands and logic
+- Write tests for new functionality
+- Might need to redesign and build new infrastructure for environment testing (Windows, macOS, REHL environments)
+- Run a few CLI commands on new supported platforms (Windows, macOS, REHL environments)
+- Add documentation
+
+#### Takeways and recommendation
+
+Ruby requires the least amount of work in terms of what is already fuctionally supported but its UX is not intuitive. As part of its deployment, the CLI requires a specific version of Ruby installed on the machine to run the CLI.
+
+Machines may not have internet access so we will need to pre-package Ruby and other dependencies and deliver as with an installer.
+
+Go CLI would require a full rewrite due to it's limited functionality, test coverage, and architectural flaws. 
+
+Python has most of the basic actions needed to interface with the Conjur server. We will need to fill in the gaps but is architecturally more stable and well tested. The Python CLI we deliver an executable so the user will not need install additional dependencies.
+
+##### Recommendation
+
+Python CLI because of its better design and it acts as a great point of entry. In terms of a model and architectural design, it is where we want to be so it is a great first step into reforming our APIs.
```

Comment:
I'd like to challenge this choice. Although the Python library might have the best interface, it suffers the same problem as Ruby in that they both require an interpreter to run on a host system. This means users need to install Python (particularly Python 3.x). Ruby is already well tested and has all the need functionality (although not the interface we'd like). In terms of speed of delivery (providing Dockerless DAP with a CLI), it feels like a clear favorite.

If we're looking at long term strategic value, I'd argue the GoLang CLI is the best option. It could take the desired interface patterns from the Python CLI.  It would be installable on nearly any chipset/OS combination with zero dependencies. It would have a tiny footprint (due to compilation).  We already have a number of highly proficient GoLang engineers (anyone who's done substantial work on Secretless ).

Reply:
>  I'd like to challenge this choice.

I want to clarify that this is a recommendation not the actual choice we are yet pursuing

> There's a link elsewhere in the doc that I posted that might be able to do the same for Ruby code

https://github.com/pmq20/ruby-packer . Tried it Thursday and now today and it is a lot less easy to get up than it seems. Sat with Gary for a good hour and a half. Nonetheless, still investing because of its potential. 

>  understand I think the timelines for this project and speed of Go development vs Ruby/Python may not make it a feasible choice

Yes. In the long run, I agree Go CLI is definitely a better decision. However, we do not have the time for a complete rewrite of functionality and testing because of this hard deadline.